### PR TITLE
[spirv] Allocate Function memory after tiling to threads

### PR DESCRIPTION
**Pull request by antiagainst**
This change fixes correctness issue due to different threads
overwriting one another's data in workgroup memory.
Bufferization after tiling to threads should use function memory,
not workgroup memory. Workgroup memory should be used if
performing bufferization before tiling to threads.
Generally we don't want allocations given they can round trip
to memory and can be slow. But at least this makes it correct.

### CODE REVIEW

**Comment by antiagainst** [comment]
This should unblock https://github.com/google/iree/pull/9103 before we can land https://github.com/google/iree/pull/8836.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**
